internal/datastore/common/revisions/optimized.go

```diff
@@ -0,0 +1,103 @@
+package revisions
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/benbjohnson/clock"
+	"github.com/rs/zerolog/log"
+	"github.com/shopspring/decimal"
+	"go.opentelemetry.io/otel"
+	"golang.org/x/sync/singleflight"
+
+	"github.com/authzed/spicedb/internal/datastore"
+)
+
+var tracer = otel.Tracer("spicedb/internal/datastore/common/revisions")
+
+// OptimizedRevisionFunction instructs the datastore to compute its own current
+// optimized revision given the specific quantization, and return for how long
+// it will remain valid.
+type OptimizedRevisionFunction func(context.Context) (rev datastore.Revision, validFor time.Duration, err error)
+
+// NewRemoteClockRevisions returns a RemoteClockRevisions for the given configuration
+func NewCachedOptimizedRevisions(maxRevisionStaleness time.Duration) *CachedOptimizedRevisions {
+	rev := atomicRevision{}
+	rev.set(validRevision{decimal.Zero, time.Time{}})
+	return &CachedOptimizedRevisions{
+		maxRevisionStaleness:  maxRevisionStaleness,
+		lastQuantizedRevision: &rev,
+		clockFn:               clock.New(),
+	}
+}
+
+// SetOptimizedRevisionFunc must be called after construction, and is the method
+// by which one specializes this helper for a specific datastore.
+func (cor *CachedOptimizedRevisions) SetOptimizedRevisionFunc(revisionFunc OptimizedRevisionFunction) {
+	cor.optimizedFunc = revisionFunc
+}
+
+func (cor *CachedOptimizedRevisions) OptimizedRevision(ctx context.Context) (datastore.Revision, error) {
+	ctx, span := tracer.Start(ctx, "OptimizedRevision")
+	defer span.End()
+
+	localNow := cor.clockFn.Now()
+	lastRevision := cor.lastQuantizedRevision.get()
+	if localNow.Before(lastRevision.validThrough) {
+		log.Debug().Time("now", localNow).Time("valid", lastRevision.validThrough).Msg("returning cached revision")
+		return lastRevision.revision, nil
+	}
+
+	lastQuantizedRevision, err, _ := cor.updateGroup.Do("", func() (interface{}, error) {
+		log.Debug().Time("now", localNow).Time("valid", lastRevision.validThrough).Msg("computing new revision")
+
+		optimized, validFor, err := cor.optimizedFunc(ctx)
```

Comment:
Add a panic here if it is not set? its slightly nicer than just an NPE

Reply:
At the cost of requiring an extra conditional every time.

Reply:
Yes, but I tend to think having panics when its an *expected* state makes sense; it would mean they forgot to call the assignment method, and that's certainly an expected bug

Reply:
I wish we had debug-only checks

Reply:
I've seen this done with build tags in https://github.com/dlsniper/debugger and in some etcd fault injection library that I can't find now. I assume the compiler removes no-op functions but I haven't checked.